### PR TITLE
feat(array): add uniqueByKey

### PR DESCRIPTION
**Pull request by blimmer**
## Description

`uniqueObjectsByKey` was in on our the private React frontends. Adding it here to fix that build.

I also added a few simple tests, since none existed in the private copy.

### CODE REVIEW

File: src/array.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@ import type { KeysOfType } from "./utils";
 declare global {
   interface Array<T> {
     unique(): Array<T>;
+    uniqueObjectsByKey(key: string): Array<T>;
```
Comment (stephenh) [inline src/array.ts:6]: Can key be typed as `keyof T`?

Comment (stephenh) [inline src/array.ts:6]: Fwiw can we rename this to uniqueByKey? I guess it can only work on objects but 🤷 I think the "by key" suffix implies that sufficiently.

File: src/array.ts [hunk 2]
```diff
@@ -144,6 +145,16 @@ Array.prototype.unique = function () {
   return [...new Set(this)];
 };
 
+/** Would be cool to allow an array of keys to make the criteria of "unique" more flexible */
+Array.prototype.uniqueObjectsByKey = function <T>(key: string) {
```
Comment (stephenh) [inline src/array.ts:149]: Wonder if `function <T extends object>` matters?

**Review by stephenh — APPROVED** [review]

File: src/array.ts (hunk 1 above)
Comment (blimmer) [reply src/array.ts]: Yep, nice improvement!

File: src/array.ts (hunk 2 above)
Comment (blimmer) [reply src/array.ts]: I played around with this a bit and it didn't seem to provide better type hints.
